lib/auth/mongocr.js

```diff
@@ -1,192 +1,54 @@
-
-  // Valid connections
-  var numberOfValidConnections = 0;
-  var errorObject = null;
-
-  // For each connection we need to authenticate
-  while (connections.length > 0) {
-    // Execute MongoCR
-    var executeMongoCR = function(connection) {
-      // Write the commmand on the connection
-      runCommand(connection, `${db}.$cmd`, { getnonce: 1 }, (err, r) => {
-        var nonce = null;
-        var key = null;
-
-        // Adjust the number of connections left
-        // Get nonce
-        if (err == null) {
-          nonce = r.result.nonce;
-          // Use node md5 generator
-          var md5 = crypto.createHash('md5');
-          // Generate keys used for authentication
-          md5.update(username + ':mongo:' + password, 'utf8');
-          var hash_password = md5.digest('hex');
-          // Final key
-          md5 = crypto.createHash('md5');
-          md5.update(nonce + username + hash_password, 'utf8');
-          key = md5.digest('hex');
-        }
-
-        // Execute command
-        // Write the commmand on the connection
-        runCommand(
-          connection,
-          `${db}.$cmd`,
-          {
-            authenticate: 1,
-            user: username,
-            nonce: nonce,
-            key: key
-          },
-          (err, r) => {
-            count = count - 1;
-
-            // If we have an error
-            if (err) {
-              errorObject = err;
-            } else if (r.result['$err']) {
-              errorObject = r.result;
-            } else if (r.result['errmsg']) {
-              errorObject = r.result;
-            } else {
-              numberOfValidConnections = numberOfValidConnections + 1;
-            }
-
-            // We have authenticated all connections
-            if (count === 0 && numberOfValidConnections > 0) {
-              // Store the auth details
-              addAuthSession(self.authStore, new AuthSession(db, username, password));
-              // Return correct authentication
-              callback(null, true);
-            } else if (count === 0) {
-              if (errorObject == null)
-                errorObject = new MongoError(f('failed to authenticate using mongocr'));
-              callback(errorObject, false);
-            }
-          }
-        );
-      });
-    };
-
-    var _execute = function(_connection) {
-      process.nextTick(function() {
-        executeMongoCR(_connection);
-      });
-    };
-
-    _execute(connections.shift());
-  }
-};
+class MongoCR extends AuthProvider {
+  /**
+   * Implementation of authentication for a single connection
+   * @override
+   */
+  _authenticateSingleConnection(sendAuthCommand, connection, credentials, callback) {
+    const username = credentials.username;
+    const password = credentials.password;
+    const source = credentials.source;
+
+    // Write the commmand on the connection
+    const getNonceCommand = this._buildCommand(`${source}.$cmd`, { getnonce: 1 });
```

Comment:
Looks like this is causing test failures because `_buildCommand` isn't a function.

Reply:
thanks, I forgot to push that change